File: src/seed.js
import defaultMindFiles from "virtual:default-mind-files";

const MIND_DIR_NAME = "my-family";

/**
 * Seed the default mind into the store if no minds exist yet.
 *
 * Writes csvs files directly to the filesystem so that mindzoo's
 * catalog.rebuild() discovers the mind on init.
 *
 * @param {object} fs - filesystem with promises API (LightningFS or node:fs)
 * @param {string} dir - root store directory ("/" for browser, store path for tauri)
 * @returns {Promise<boolean>} true if seeded, false if store already had data
 */
export async function seedDefaultMind(fs, dir) {
  // check if any directories exist (besides "root" which rebuild creates)
  let entries;

  try {
    entries = await fs.promises.readdir(dir);
  } catch {
    entries = [];
  }

  const hasMinds = entries.some((e) => e !== "root");

  if (hasMinds) return false;

  const mindDir = `${dir}/${MIND_DIR_NAME}`;

  // create the mind root dir
  try {
    await fs.promises.mkdir(mindDir);
  } catch {
    // already exists
  }

  // collect all unique subdirectories we need to create
  const dirs = new Set();

  for (const relPath of Object.keys(defaultMindFiles)) {
    const parts = relPath.split("/");

    // build each intermediate directory path
    for (let i = 1; i <= parts.length - 1; i++) {
      dirs.add(parts.slice(0, i).join("/"));
    }
  }

  // create subdirectories (sorted ensures parents come first)
  for (const d of [...dirs].sort()) {
    try {
      await fs.promises.mkdir(`${mindDir}/${d}`);
    } catch {
      // already exists
    }
  }

  // write all files
  for (const [relPath, content] of Object.entries(defaultMindFiles)) {
    await fs.promises.writeFile(`${mindDir}/${relPath}`, content);
  }

  return true;
}

File: src-tauri/src/seed.rs
use std::fs;
use std::path::Path;

include!(concat!(env!("OUT_DIR"), "/default_mind_data.rs"));

const MIND_DIR_NAME: &str = "my-family";

/// Seed the default mind into the store if no minds exist yet.
///
/// Writes csvs files directly to the filesystem so that Mindzoo::new()
/// discovers the mind on rebuild.
///
/// Returns true if seeded, false if store already had data.
pub fn seed_default_mind(store_dir: &Path) -> std::io::Result<bool> {
    eprintln!("[seed] store_dir={} exists={}", store_dir.display(), store_dir.exists());
    eprintln!("[seed] DEFAULT_MIND_FILES has {} entries", DEFAULT_MIND_FILES.len());

    // check if any directories exist (besides "root")
    let has_minds = fs::read_dir(store_dir)?
        .filter_map(|e| e.ok())
        .filter(|e| e.file_type().map(|ft| ft.is_dir()).unwrap_or(false))
        .any(|e| {
            let name = e.file_name();
            eprintln!("[seed] found dir: {:?}", name);
            name != "root"
        });

    eprintln!("[seed] has_minds={}", has_minds);

    if has_minds {
        return Ok(false);
    }

    let mind_dir = store_dir.join(MIND_DIR_NAME);
    eprintln!("[seed] writing to {}", mind_dir.display());

    for (rel_path, content) in DEFAULT_MIND_FILES {
        let full_path = mind_dir.join(rel_path);

        if let Some(parent) = full_path.parent() {
            fs::create_dir_all(parent)?;
        }

        fs::write(&full_path, content)?;
    }

    eprintln!("[seed] done, wrote {} files", DEFAULT_MIND_FILES.len());

    Ok(true)
}
